[GEOS-11421] Match webui workspace admin functionality in the REST API

groldan opened the pull request:
[![GEOS-11421](https://badgen.net/badge/JIRA/GEOS-11421/0052CC)](https://osgeo-org.atlassian.net/browse/GEOS-11421) [<img width="16" alt="Powered by Pull Request Badge" src="https://user-images.githubusercontent.com/1393946/111216524-d2bb8e00-85d4-11eb-821b-ed4c00989c02.png">](https://pullrequestbadge.com/?utm_medium=github&utm_source=geoserver&utm_campaign=badge_info)<!-- PR-BADGE: PLEASE DO NOT REMOVE THIS COMMENT -->  

## Summary

Extends workspace admin access control to the REST API, matching the permissions workspace administrators already have in the web interface. The default `rest.properties` restricts all REST endpoints to `ROLE_ADMINISTRATOR`, and while it could be edited to grant any user role access to specific endpoints, it has no concept of workspace scoping. This change adds a complementary configuration (`rest.workspaceadmin.properties`) that works alongside `rest.properties` to grant REST access to any user identified as a workspace administrator, with visibility scoped to their administrable workspaces.

## Security layer

- `WorkspaceAdminAuthorizer`: central authorization component shared by web UI and REST API, replacing the previous web-only implementation in `WorkspaceAdminComponentAuthorizer`. Determines workspace admin status with request-scoped caching for performance.
- `WorkspaceAdminAuthorizationManager`: added to the Spring Security filter chain (`AffirmativeBased`, alongside existing `AuthenticatedAuthorizationManager` and `RoleAuthorizationManager`). Returns `ACCESS_GRANTED` when a matching workspace admin rule is found, `ACCESS_ABSTAIN` otherwise — letting other voters handle non-workspace-admin requests.
- `WorkspaceAdminRestfulDefinitionSource` and `WorkspaceAdminRESTAccessRuleDAO`: load URL pattern-based access rules from `rest.workspaceadmin.properties` (initialized from a built-in template on first startup).
- `RESTfulDefinitionSourceProxy`: combines standard REST security rules (`rest.properties`) with workspace admin rules in a single `SecurityMetadataSource`.

## Access rules (`rest.workspaceadmin.properties`)

Rule order matters — first match wins. Unmatched endpoints fall through to `rest.properties` (admin-only by default).

| Endpoint pattern | Access | Notes |
|---|---|---|
| `/rest/workspaces`, `/rest/namespaces` | read | Collection listings (filtered to administrable workspaces) |
| `/rest/workspaces/{workspace}` | read + PUT | Update properties, no rename or delete |
| `/rest/workspaces/{workspace}/**` | read/write | Full access to sub-resources (datastores, styles, layergroups, etc.) |
| `/rest/layers/**` | read/write | Filtered to administrable workspaces |
| `/rest/styles/**`, `/rest/templates/**` | read | Global styles/templates are read-only |
| `/rest/resource/workspaces/{workspace}/**` | read/write | Resource browser, scoped to administrable workspaces |
| `/rest/resource/styles/**` | read | Global style resources, read-only |
| `/rest/services/*/workspaces/{workspace}/**` | read/write | Per-workspace service config (WFS, WMS, etc.) |
| `/rest/crs/**`, `/rest/fonts/**` | read | CRS lookups and fonts |
| `/rest/services/*/settings` | *(no rule)* | Falls through to admin-only |

`rest.workspaceadmin.properties` opens the door to endpoints, `layers.properties` determines what's behind the door — the catalog security layer filters which workspaces and resources are actually visible based on the admin access rules.

## Resource browser security

- `SecureResourceStore`: wraps the data directory `ResourceStore`, filtering access through `WorkspaceAdminResourceFilter` based on workspace admin privileges.
- `SecuredResource`: decorator that enforces read/write checks through the store for all operations including `file()` and `dir()`. Throws `RestException` with proper HTTP status (404 if not visible, 403 if read-only).

## Controller fixes

- `WorkspaceController` / `NamespaceController`: deny rename attempts by workspace admins (403).
- `ServiceSettingsController`: null-check workspace after catalog lookup to prevent fallthrough to global settings when the workspace is hidden.
- `TemplateController`: validate workspace, store, and resource existence before operations.
- `LocalSettingsController`: validate workspace existence to prevent NPE.
- `IndexController`: filter REST API index links based on workspace admin access rules.

## Documentation

- New **Workspace Administration** (`doc/en/user/security/workspaceadmin/index.md`) landing page explaining the concept and the three configuration layers (`layers.properties`, `rest.properties`, `rest.workspaceadmin.properties`).
- REST access rules reference page under the new section.
- New **tutorial** (doc/en/user/security/tutorials/workspaceadmin/index.md) walking through setup and REST API exploration using the default data directory.
- Cross-links added from layer security, sandbox, production config, and data admin pages.

## Tests

- Unit tests: `WorkspaceAdminAuthorizerTest`, `WorkspaceAdminAuthorizationManagerTest`, `WorkspaceAdminRESTAccessRuleDAOTest`, `WorkspaceAdminRestAccessRuleTest`, `WorkspaceAdminRestfulDefinitionSourceTest`, `RESTfulDefinitionSourceProxyTest`, `SecuredResourceTest`, `SecureResourceStoreTest`, `WorkspaceAdminResourceFilterTest`
- Integration tests: `WorkspaceAdminRestIntegrationTest`, `ResourceControllerWorkspaceAdminTest`, `WMSSettingsWorkspaceAdminIntegrationTest`, `WFSSettingsWorkspaceAdminIntegrationTest`

<!--Include a few sentences describing the overall goals for this Pull Request-->
  
<!-- Please help our volunteers reviewing this PR by completing the following items. 
Ask in a comment if you have troubles with any of them. -->

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [ ] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

<!--Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or not applicable.-->

aaime commented:
Just looking at the comment above and thinking out loud... what about the ``/layers`` and/or ``/workspaces/{workspaceName}/layers`` resources? 
A workspace admin can also alter publishing details like styles, HTTP caching, and so on.

aaime commented:
I see activity after a long hiatus, but the PR is still marked as Draft. So not ready yet for review, yes?

groldan commented:
> I see activity after a long hiatus, but the PR is still marked as Draft. So not ready yet for review, yes?

Hi Andrea, that was the fix merge conflicts/rebase after the palantir reformatting I guess. Now I'm actively working on getting it ready, mostly in terms of documentation and additional tests

vuilleumierc commented on src/main/src/main/java/org/geoserver/security/RESTfulDefinitionSourceProxy.java:
```diff
@@ -0,0 +1,84 @@
+/* (c) 2014 Open Source Geospatial Foundation - all rights reserved
```
Wrong year in copyright - please also check other files

vuilleumierc commented on src/main/src/test/java/org/geoserver/security/workspaceadmin/WorkspaceAdminAuthorizationManagerTest.java:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    @Test
+    public void testCheckWhenMatchingRuleAndWorkspaceAdmin() {
```
Would it make sense to also test the case of a workspace admin but there's no matching rule?

vuilleumierc commented on src/restconfig/src/test/java/org/geoserver/rest/resources/SecureResourceStoreTest.java:
```diff
@@ -0,0 +1,196 @@
+        // Should not delegate to underlying store when not allowed
+        assertFalse(secureStore.remove("workspaces/sf"));
+
+        // Note: verify that remove was not called on delegate
```
a TODO is left

vuilleumierc reviewed (approved):
LGTM and tested locally, I only have minor comments

jodygarnett commented:
@vuilleumierc and @groldan This one is open and appears good to go? 

Do you want this in for the 2.27.0 release?

groldan commented:
Hi @jodygarnett, this is a contribution from the GeoServer Cloud project. We're not in a rush to have it merged and I'd rather have it reviewed by someone else too. Cecile has been using it for months for a customer project. This PR is pretty complete but I'm waiting to finish up a series of integration tests to complement the unit tests in here.
That is to say, I don't want this to delay the 2.27 release, nor rush it up, since it's a pretty critical functionality affecting security.

groldan committed:
[GEOS-11421] Match webui workspace admin functionality in the REST API

Extend workspace admin access control to the REST API, ensuring workspace
administrators have the same fine-grained permissions via REST as they do
through the web interface.

- WorkspaceAdminAuthorizer: central authorization component shared by
  web UI and REST API to check if an authenticated user is a workspace
  administrator and can access a given resource (workspace or URI).
- WorkspaceAdminAuthorizationManager: plugs into the Spring Security
  filter chain alongside existing role/auth voters
- rest.workspaceadmin.properties: configurable URL pattern-based access
  rules (read-only for global resources, full access within administrable
  workspaces, unmatched endpoints fall through to admin-only)
- SecureResourceStore/SecuredResource: enforce REST resource access
  control based on workspace admin privileges
- ServiceSettingsController: fix null workspace fallthrough to global settings
- WorkspaceController/NamespaceController: deny rename by workspace admins
- Documentation: workspace admin landing page, REST reference, and tutorial

on-behalf-of: @camptocamp <info@camptocamp.com>

groldan commented:
This is finally ready for review. Added full test coverage and documentation

groldan commented:
Rebased on top of the latest main branch.
@aaime @jodygarnett I know it's a large PR, but if someone in the PSC can review that'd be great.
Note again this functionality is being used for years now in GeoServer Cloud production deployments, and we're trying to push it upstream within the ongoing effort to prescind of the geoserver fork we use to apply required patched on top of the geoserver branch.
TIA,
Gabe

aaime commented:
This is too big to review alongside other PRs, but I made a first pass using an AI trained to prioritize the same aspects I inspect when performing manual reviews. Call it a first pass:

-----

  Overall assessment

  This is a well-structured, security-conscious feature with strong integration-test coverage and genuinely good user documentation. The architecture (abstain-only voter that can only widen access, two-layer
  "door vs behind-door" model, request-scoped caching, deferred resource checks) is sound and the dangerous path — filesystem traversal — is properly defended. The main blemishes are not logic bugs but
  discipline issues: a comparator/equals inconsistency that can silently drop a security rule, one undocumented behavior change, and a heavy layer of AI-style boilerplate javadoc/comments that clashes with the
  repo's terse style and should be trimmed before merge.

  Three to address before merge:
  1. Finding nr. 1 — make WorkspaceAdminRestAccessRule.compareTo consistent with equality so no rule can be silently dropped on a priority collision.
  2. Finding nr. 2 — confirm and document the ServiceSettingsController 404 behavior change (and update REST API docs).
  3. Finding nr. 3,4,5 — strip the restating javadoc/inline comments and fix the stale/invented rule lists, per house style.

  Verdict: Request changes — no blocking defects, but nr 1 (silent rule loss) and the documentation/comment cleanup should land before merge.

----

Given this, I'm not very inclined to get sucked into Shakespearean style docs. Can you clean up the documentation/comments so that it's quicker to understand, both for the poor reviewer and the random developer that wil stumble into the code later? Then I'll do an actual manual pass.

groldan commented:
Hi @aaime, thanks for the review; addressed the three points: 
- `compareTo` is consistent with equals
- `ServiceSettingsController`'s `404` behavior change documented in `doc/en/user/rest/api/services.md`. Added
```markdown
### Exceptions

| Exception                                              | Status code |
|--------------------------------------------------------|-------------|
| GET, PUT or DELETE for a workspace that does not exist | 404         |
```
which covers both whether the workspace exists or is not visible (for a workspace admin it's the same)

-  agreed on the AI-style boilerplate javadoc/comments. They're now trimmed down with the invented rule lists removed; and the bundled template is now the single source of truth.

jurgispods commented:
Hi @groldan — we've been running this PR's code in production for several months (it ships in the
geoserver-cloud 2.28.4 Docker images), on a
~34k-layer / ~1k-workspace catalog. It works well. We did hit one issue in
`WorkspaceAdminResourceFilter` that looks unintended, and since the PR is still open it seems better
to report it here.

### `workspaceResourceAccess()` can never return `READ`

```java
private ResourceAccess workspaceResourceAccess(Authentication authentication, final String workspace) {
    WorkspaceAccessLimits wsAccessLimits = authorizer.getWorkspaceAccessLimits(authentication, workspace);
    boolean adminable = wsAccessLimits != null && wsAccessLimits.isAdminable();
    return adminable ? ResourceAccess.WRITE : ResourceAccess.NONE;
}
```

`WorkspaceAccessLimits.isReadable()` is never consulted, so any non-workspace-admin gets `NONE` for
`workspaces/{ws}/**` even when the catalog says the workspace is readable by everyone (`*.*.r=*` in
`layers.properties`). `SecureResourceStore` maps `NONE` to `Resource.Type.UNDEFINED`, and under
`mode=HIDE` that surfaces as a 404 from `/rest/resource/workspaces/{ws}/**`. This is indistinguishable from
"the file isn't there", which made it take a while to track down.

Concretely it broke anonymous reads of style assets - SVG icons referenced from SLDs and served from the resource store - which worked before we migrated to GeoServer Cloud, since standard Geoserver has no
equivalent filter. Note this is independent of `security/rest.properties`: an explicit
`IS_AUTHENTICATED_ANONYMOUSLY` grant for that exact path passes the `AuthorizationFilter` and then
still 404s here.

The sibling method already models what we'd expect for this case:

```java
private ResourceAccess noWorkspaceResourceAccess(Authentication authentication, String path) {
    boolean readable = isAllowedCollectionPath(path) && authorizer.isWorkspaceAdmin(authentication);
    return readable ? ResourceAccess.READ : ResourceAccess.NONE;
}
```

so both `ResourceAccess.READ` and `isReadable()` already exist and are simply unused for the
workspace path.

### Suggested change

```java
 private ResourceAccess workspaceResourceAccess(Authentication authentication, final String workspace) {
     WorkspaceAccessLimits wsAccessLimits = authorizer.getWorkspaceAccessLimits(authentication, workspace);
-    boolean adminable = wsAccessLimits != null && wsAccessLimits.isAdminable();
-    return adminable ? ResourceAccess.WRITE : ResourceAccess.NONE;
+    if (wsAccessLimits == null) {
+        return ResourceAccess.WRITE;
+    }
+    if (wsAccessLimits.isAdminable()) {
+        return ResourceAccess.WRITE;
+    }
+    if (wsAccessLimits.isReadable()) {
+        return ResourceAccess.READ;
+    }
+    return ResourceAccess.NONE;
 }
```

Tested locally against the 2.28.4.0 images with the patch applied, `layers.properties`
granting `*.*.r=*`:

| request                                                     | before | after |
|-------------------------------------------------------------|--------|-------|
| admin `GET /rest/resource/workspaces/{ws}/styles/test.svg`     | 200    | 200   |
| anonymous `GET /rest/resource/workspaces/{ws}/styles/test.svg` | 404    | 200   |
| anonymous `PUT`/`DELETE` on the same resource                  | 401    | 401   |
| anonymous `GET /rest/workspaces/{ws}/styles.json`              | 401    | 401   |
| anonymous `GET /rest/workspaces.json`                          | 401    | 401   |

i.e. only the read-only case changes; write still requires `isAdminable()`, and the catalog endpoints
are untouched.

### One judgement call I'd like your opinion on: the `null` branch

`DefaultResourceAccessManager.getAccessLimits(Authentication, WorkspaceInfo)` returns `null` rather
than a populated `WorkspaceAccessLimits` when the user is readable **and** writable but not adminable
and `AdminRequest.get() == null` - its own comment reads *"It is not an admin request, read+write
means full access"*. The current `wsAccessLimits != null && isAdminable()` test reads that `null` as
"no access", which inverts its meaning. We hit this in a second round: a JWT-authenticated, non-admin
service account with workspace write access still got 404s after the `isReadable()` case alone was
fixed.

Mapping `null` to `WRITE` follows that contract literally, and is what we run. But `READ` would be the
conservative choice and would also have fixed our symptom, so I'd rather you decide.

If it helps, I'm happy to open a PR against `GEOS-11421_admin_workspaces_rest_api` including
tests.
